src/github.com/getlantern/flashlight/client/handler.go

```diff
@@ -93,35 +101,50 @@ func (client *Client) targetQOS(req *http.Request) int {
 // pipeData pipes data between the client and proxy connections.  It's also
 // responsible for responding to the initial CONNECT request with a 200 OK.
 func pipeData(clientConn net.Conn, connOut net.Conn, req *http.Request) {
-
-	var wg sync.WaitGroup
-	wg.Add(1)
-	// Start piping from client to proxy
-	go func() {
-		io.Copy(connOut, clientConn)
-		// Force closing if EOF at the request half or error encountered.
-		// A bit arbitrary, but it's rather rare now to use half closing
-		// as a way to notify server. Most application closes both connections
-		// after completed send / receive so that won't cause problem.
-		wg.Wait()
-		clientConn.Close()
-		connOut.Close()
-	}()
-
 	// Respond OK
 	err := respondOK(clientConn, req)
-	wg.Done()
 	if err != nil {
 		log.Errorf("Unable to respond OK: %s", err)
 		return
 	}
 
+	// Make sure of closing connections only once
+	var closeOnce sync.Once
+
+	// Force closing if EOF at the request half or error encountered.
+	// A bit arbitrary, but it's rather rare now to use half closing
+	// as a way to notify server. Most application closes both connections
+	// after completed send / receive so that won't cause problem.
+	closeConns := func() {
+		if err := clientConn.Close(); err != nil {
+			log.Debugf("Error closing the out connection: %s", err)
+		}
+		if err := connOut.Close(); err != nil {
+			log.Debugf("Error closing the client connection: %s", err)
+		}
+	}
+
+	// Start piping from client to proxy
+	go func() {
+		if _, err := io.Copy(connOut, clientConn); err != nil {
+			log.Debugf("Error piping data from client to proxy: %s", err)
+		}
+		closeOnce.Do(closeConns)
+	}()
+
 	// Then start coyping from proxy to client
-	io.Copy(clientConn, connOut)
+	if _, err := io.Copy(clientConn, connOut); err != nil {
+		log.Debugf("Error piping data from proxy to client: %s", err)
+	}
+	closeOnce.Do(closeConns)
```

Comment:
One comment about this line too. After testing it with extra error handling I realized we were closing twice. This code would take care of that (pending in another branch):

```

// pipeData pipes data between the client and proxy connections.
func pipeData(clientConn net.Conn, connOut net.Conn, req *http.Request) {
	// Wait for both directions to be done
	done := make(chan bool)

	// Start piping from client to proxy
	go func() {
		if _, err := io.Copy(connOut, clientConn); err != nil {
			log.Debugf("Error piping data from client to proxy (%v): %s", err, connOut)
		}
		done <- true
	}()

	// Then start copying from proxy to client
	if _, err := io.Copy(clientConn, connOut); err != nil {
		log.Debugf("Error piping data from proxy to client (%v): %s", err, connOut)
	}
	<-done
}
```